osclib/core.py

```diff
@@ -431,7 +431,11 @@ def repository_path_expand(apiurl, project, repo):
 
     # ...which is expanded recursively
     if len(paths) > 0:
-        repos += repository_path_expand(apiurl, paths[-1].get('project', project), paths[-1].get('repository'))
+        p_project = paths[-1].get('project', project)
+        p_repository = paths[-1].get('repository')
+        if (p_project, p_repository) not in visited_repos:
+            visited_repos.add((p_project, p_repository))
```

Comment:
I think you're falling into a nasty python trap here: defaulted arguments are persistent and mutable!
The next time this is called, the defaulted `repository_path_expand` has the content from last time...

```
def a(b, c=set()):
        c.add(b)
        return c

print(a("foo"))
print(a("bar"))
```

```
{'foo'}
{'foo', 'bar'}
```

Reply:
I'm using the =None way of handling defaulted arguments.